// @volar/vscode-vue-languageservice/out/services/documentSymbol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const node_1 = require("vscode-languageserver/node");
const languageServices = require("../utils/languageServices");
const shared_1 = require("@volar/shared");
function register({ sourceFiles, tsLanguageService }) {
    return (document) => {
        const sourceFile = sourceFiles.get(document.uri);
        if (!sourceFile)
            return;
        const vueResult = getVueResult(sourceFile);
        const tsResult = getTsResult(sourceFile);
        const htmlResult = getHtmlResult(sourceFile);
        const cssResult = getCssResult(sourceFile);
        // TODO: pug
        return [
            ...vueResult,
            ...tsResult,
            ...htmlResult,
            ...cssResult,
        ];
        function getVueResult(sourceFile) {
            const result = [];
            const desc = sourceFile.getDescriptor();
            if (desc.template) {
                result.push({
                    name: '<template>',
                    kind: node_1.SymbolKind.Module,
                    location: node_1.Location.create(document.uri, node_1.Range.create(document.positionAt(desc.template.loc.start), document.positionAt(desc.template.loc.end))),
                });
            }
            if (desc.script) {
                result.push({
                    name: '<script>',
                    kind: node_1.SymbolKind.Module,
                    location: node_1.Location.create(document.uri, node_1.Range.create(document.positionAt(desc.script.loc.start), document.positionAt(desc.script.loc.end))),
                });
            }
            if (desc.scriptSetup) {
                result.push({
                    name: '<script setup>',
                    kind: node_1.SymbolKind.Module,
                    location: node_1.Location.create(document.uri, node_1.Range.create(document.positionAt(desc.scriptSetup.loc.start), document.positionAt(desc.scriptSetup.loc.end))),
                });
            }
            for (const style of desc.styles) {
                result.push({
                    name: `<${['style', style.scoped ? 'scoped' : undefined, style.module ? 'module' : undefined].filter(shared_1.notEmpty).join(' ')}>`,
                    kind: node_1.SymbolKind.Module,
                    location: node_1.Location.create(document.uri, node_1.Range.create(document.positionAt(style.loc.start), document.positionAt(style.loc.end))),
                });
            }
            for (const customBlock of desc.customBlocks) {
                result.push({
                    name: `<${customBlock.type}>`,
                    kind: node_1.SymbolKind.Module,
                    location: node_1.Location.create(document.uri, node_1.Range.create(document.positionAt(customBlock.loc.start), document.positionAt(customBlock.loc.end))),
                });
            }
            return result;
        }
        function getTsResult(sourceFile) {
            const result = [];
            const map = new Map();
            for (const sourceMap of sourceFile.getTsSourceMaps()) {
                if (!sourceMap.capabilities.documentSymbol)
                    continue;
                let symbols = tsLanguageService.findWorkspaceSymbols(sourceMap.mappedDocument.uri);
                for (const s of symbols) {
                    const vueRange = sourceMap.getSourceRange(s.location.range.start, s.location.range.end);
                    if (vueRange) {
                        map.set(`${sourceMap.mappedDocument.offsetAt(s.location.range.start)}:${sourceMap.mappedDocument.offsetAt(s.location.range.end)}:${s.kind}:${s.name}`, {
                            ...s,
                            location: node_1.Location.create(document.uri, vueRange),
                        });
                    }
                }
            }
            for (const info of map.values()) {
                result.push(info);
            }
            return result;
        }
        function getHtmlResult(sourceFile) {
            const result = [];
            const sourceMaps = sourceFile.getHtmlSourceMaps();
            for (const sourceMap of sourceMaps) {
                let symbols = languageServices.html.findDocumentSymbols(sourceMap.mappedDocument, sourceMap.htmlDocument);
                if (!symbols)
                    continue;
                for (const s of symbols) {
                    const vueRange = sourceMap.getSourceRange(s.location.range.start, s.location.range.end);
                    if (vueRange) {
                        result.push({
                            ...s,
                            location: node_1.Location.create(document.uri, vueRange),
                        });
                    }
                }
            }
            return result;
        }
        function getCssResult(sourceFile) {
            const result = [];
            const sourceMaps = sourceFile.getCssSourceMaps();
            for (const sourceMap of sourceMaps) {
                const cssLanguageService = languageServices.getCssLanguageService(sourceMap.mappedDocument.languageId);
                if (!cssLanguageService || !sourceMap.stylesheet)
                    continue;
                let symbols = cssLanguageService.findDocumentSymbols(sourceMap.mappedDocument, sourceMap.stylesheet);
                if (!symbols)
                    continue;
                for (const s of symbols) {
                    const vueRange = sourceMap.getSourceRange(s.location.range.start, s.location.range.end);
                    if (vueRange) {
                        result.push({
                            ...s,
                            location: node_1.Location.create(document.uri, vueRange),
                        });
                    }
                }
            }
            return result;
        }
    };
}
exports.register = register;
