// @volar/vscode-vue-languageservice/out/virtuals/script.raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useScriptFormat = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const shared_1 = require("@volar/shared");
const reactivity_1 = require("@vue/reactivity");
const SourceMaps = require("../utils/sourceMaps");
function useScriptFormat(getUnreactiveDoc, script) {
    let version = 0;
    const textDocument = reactivity_1.computed(() => {
        if (script.value) {
            const vueDoc = getUnreactiveDoc();
            const lang = shared_1.getValidScriptSyntax(script.value.lang);
            const uri = `${vueDoc.uri}.${lang}`;
            return vscode_languageserver_textdocument_1.TextDocument.create(uri, shared_1.syntaxToLanguageId(lang), version++, script.value.content);
        }
    });
    const sourceMap = reactivity_1.computed(() => {
        if (textDocument.value && script.value) {
            const vueDoc = getUnreactiveDoc();
            const sourceMap = new SourceMaps.TsSourceMap(vueDoc, textDocument.value, false, {
                foldingRanges: true,
                formatting: true,
                documentSymbol: false,
                codeActions: false,
                organizeImports: false,
            });
            sourceMap.add({
                data: {
                    vueTag: 'script',
                    capabilities: {
                        formatting: true,
                        foldingRanges: true,
                    },
                },
                mode: SourceMaps.Mode.Offset,
                sourceRange: {
                    start: script.value.loc.start,
                    end: script.value.loc.end,
                },
                mappedRange: {
                    start: 0,
                    end: script.value.loc.end - script.value.loc.start,
                },
            });
            return sourceMap;
        }
    });
    return {
        textDocument,
        sourceMap,
    };
}
exports.useScriptFormat = useScriptFormat;
