// @volar/vscode-vue-languageservice/out/languageService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLanguageService = exports.createNoStateLanguageService = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const shared_1 = require("@volar/shared");
const sourceFile_1 = require("./sourceFile");
const global_1 = require("./virtuals/global");
const reactivity_1 = require("@vue/reactivity");
const upath = require("upath");
const ts2 = require("@volar/vscode-typescript-languageservice");
const languageServices = require("./utils/languageServices");
const mapper_1 = require("./utils/mapper");
const tsPluginApis = require("./tsPluginApis");
const tsProgramApis = require("./tsProgramApis");
// vue services
const completions = require("./services/completion");
const completionResolve = require("./services/completionResolve");
const autoClose = require("./services/autoClose");
const refAutoClose = require("./services/refAutoClose");
const hover = require("./services/hover");
const diagnostics = require("./services/diagnostics");
const formatting = require("./services/formatting");
const definitions = require("./services/definition");
const references = require("./services/references");
const rename = require("./services/rename");
const codeActions = require("./services/codeAction");
const codeActionResolve = require("./services/codeActionResolve");
const documentHighlight = require("./services/documentHighlight");
const documentSymbol = require("./services/documentSymbol");
const documentLink = require("./services/documentLinks");
const documentColor = require("./services/documentColor");
const selectionRanges = require("./services/selectionRanges");
const signatureHelp = require("./services/signatureHelp");
const colorPresentations = require("./services/colorPresentation");
const semanticTokens = require("./services/semanticTokens");
const foldingRanges = require("./services/foldingRanges");
const codeLens = require("./services/codeLens");
const codeLensResolve = require("./services/codeLensResolve");
const executeCommand = require("./services/executeCommand");
const callHierarchy = require("./services/callHierarchy");
const linkedEditingRanges = require("./services/linkedEditingRange");
const d3 = require("./services/d3");
const shared_2 = require("@volar/shared");
function createNoStateLanguageService({ typescript }) {
    const cache = new Map();
    const options = {
        ts: typescript,
        getHtmlDocument,
    };
    return {
        doFormatting: formatting.register(options),
        getFoldingRanges: foldingRanges.register(options),
        doAutoClose: autoClose.register(options),
        findLinkedEditingRanges: linkedEditingRanges.register(options),
    };
    function getHtmlDocument(document) {
        const _cache = cache.get(document.uri);
        if (_cache) {
            const [cacheVersion, cacheHtmlDoc] = _cache;
            if (cacheVersion === document.version) {
                return cacheHtmlDoc;
            }
        }
        const htmlDoc = languageServices.html.parseHTMLDocument(document);
        cache.set(document.uri, [document.version, htmlDoc]);
        return htmlDoc;
    }
}
exports.createNoStateLanguageService = createNoStateLanguageService;
function createLanguageService(vueHost, { typescript }, onUpdate, isTsPlugin = false) {
    let vueProjectVersion;
    let lastScriptVersions = new Map();
    let tsProjectVersion = 0;
    let tsProjectVersionWithoutTemplate = 0;
    let lastCompletionUpdateVersion = -1;
    let shouldSendUpdate = true;
    const documents = new shared_2.UriMap();
    const sourceFiles = new shared_2.UriMap();
    const templateScriptUpdateUris = new Set();
    const tsLanguageServiceHost = createTsLanguageServiceHost();
    const tsLanguageService = ts2.createLanguageService(tsLanguageServiceHost, typescript);
    const globalDoc = global_1.getGlobalDoc(vueHost.getCurrentDirectory());
    const mapper = mapper_1.createMapper(sourceFiles, tsLanguageService, getTextDocument);
    const options = {
        ts: typescript,
        sourceFiles,
        tsLanguageService,
        vueHost,
        mapper,
    };
    const _callHierarchy = callHierarchy.register(options);
    const findDefinition = definitions.register(options);
    const doRename = rename.register(options);
    // ts plugin proxy
    const _tsPluginApis = tsPluginApis.register(options);
    const tsPlugin = {
        getSemanticDiagnostics: apiHook(tsLanguageService.raw.getSemanticDiagnostics, false),
        getEncodedSemanticClassifications: apiHook(tsLanguageService.raw.getEncodedSemanticClassifications, false),
        getCompletionsAtPosition: apiHook(_tsPluginApis.getCompletionsAtPosition, false),
        getCompletionEntryDetails: apiHook(tsLanguageService.raw.getCompletionEntryDetails, false),
        getCompletionEntrySymbol: apiHook(tsLanguageService.raw.getCompletionEntrySymbol, false),
        getQuickInfoAtPosition: apiHook(tsLanguageService.raw.getQuickInfoAtPosition, false),
        getSignatureHelpItems: apiHook(tsLanguageService.raw.getSignatureHelpItems, false),
        getRenameInfo: apiHook(tsLanguageService.raw.getRenameInfo, false),
        findRenameLocations: apiHook(_tsPluginApis.findRenameLocations, true),
        getDefinitionAtPosition: apiHook(_tsPluginApis.getDefinitionAtPosition, false),
        getDefinitionAndBoundSpan: apiHook(_tsPluginApis.getDefinitionAndBoundSpan, false),
        getTypeDefinitionAtPosition: apiHook(_tsPluginApis.getTypeDefinitionAtPosition, false),
        getImplementationAtPosition: apiHook(_tsPluginApis.getImplementationAtPosition, false),
        getReferencesAtPosition: apiHook(_tsPluginApis.getReferencesAtPosition, true),
        findReferences: apiHook(_tsPluginApis.findReferences, true),
        // TODO: now is handle by vue server
        // prepareCallHierarchy: apiHook(tsLanguageService.rawLs.prepareCallHierarchy, false),
        // provideCallHierarchyIncomingCalls: apiHook(tsLanguageService.rawLs.provideCallHierarchyIncomingCalls, false),
        // provideCallHierarchyOutgoingCalls: apiHook(tsLanguageService.rawLs.provideCallHierarchyOutgoingCalls, false),
        // getEditsForFileRename: apiHook(tsLanguageService.rawLs.getEditsForFileRename, false),
        // TODO
        // getCodeFixesAtPosition: apiHook(tsLanguageService.rawLs.getCodeFixesAtPosition, false),
        // getCombinedCodeFix: apiHook(tsLanguageService.rawLs.getCombinedCodeFix, false),
        // applyCodeActionCommand: apiHook(tsLanguageService.rawLs.applyCodeActionCommand, false),
        // getApplicableRefactors: apiHook(tsLanguageService.rawLs.getApplicableRefactors, false),
        // getEditsForRefactor: apiHook(tsLanguageService.rawLs.getEditsForRefactor, false),
    };
    // ts program proxy
    const tsProgram = tsLanguageService.raw.getProgram();
    if (!tsProgram)
        throw '!tsProgram';
    const tsProgramApis_2 = tsProgramApis.register(options);
    const tsProgramApis_3 = {
        emit: apiHook(tsProgramApis_2.emit),
        getRootFileNames: apiHook(tsProgramApis_2.getRootFileNames),
        getSemanticDiagnostics: apiHook(tsProgramApis_2.getSemanticDiagnostics),
        getSyntacticDiagnostics: apiHook(tsProgramApis_2.getSyntacticDiagnostics),
        getGlobalDiagnostics: apiHook(tsProgramApis_2.getGlobalDiagnostics),
    };
    const tsProgramProxy = new Proxy(tsProgram, {
        get: (target, property) => {
            return tsProgramApis_3[property] || target[property];
        },
    });
    return {
        rootPath: vueHost.getCurrentDirectory(),
        tsPlugin,
        tsProgramProxy,
        update,
        getTextDocument,
        checkProject: apiHook(() => {
            const vueImportErrors = tsLanguageService.doValidation(globalDoc.uri, { semantic: true });
            return !vueImportErrors.find(error => error.code === 2322); // Type 'false' is not assignable to type 'true'.ts(2322)
        }),
        getTsService: () => tsLanguageService,
        getGlobalDocs: () => [globalDoc],
        getSourceFile: apiHook(getSourceFile),
        getAllSourceFiles: apiHook(getAllSourceFiles),
        doValidation: apiHook(diagnostics.register(options)),
        findDefinition: apiHook(findDefinition.on),
        findReferences: apiHook(references.register(options)),
        findTypeDefinition: apiHook(findDefinition.onType),
        callHierarchy: {
            onPrepare: apiHook(_callHierarchy.onPrepare),
            onIncomingCalls: apiHook(_callHierarchy.onIncomingCalls),
            onOutgoingCalls: apiHook(_callHierarchy.onOutgoingCalls),
        },
        rename: {
            onPrepare: apiHook(doRename.onPrepare),
            doRename: apiHook(doRename.doRename),
            onRenameFile: apiHook(doRename.onRenameFile, false),
        },
        getSemanticTokens: apiHook(semanticTokens.register(options)),
        getD3: apiHook(d3.register(options)),
        doHover: apiHook(hover.register(options), getShouldUpdateTemplateScript),
        doComplete: apiHook(completions.register(options), getShouldUpdateTemplateScript),
        getCodeActions: apiHook(codeActions.register(options), false),
        doCodeActionResolve: apiHook(codeActionResolve.register(options), false),
        doExecuteCommand: apiHook(executeCommand.register(options), false),
        doCompletionResolve: apiHook(completionResolve.register(options), false),
        doCodeLensResolve: apiHook(codeLensResolve.register(options), false),
        getSignatureHelp: apiHook(signatureHelp.register(options), false),
        getSelectionRanges: apiHook(selectionRanges.register(options), false),
        getColorPresentations: apiHook(colorPresentations.register(options), false),
        getCodeLens: apiHook(codeLens.register(options), false),
        findDocumentHighlights: apiHook(documentHighlight.register(options), false),
        findDocumentSymbols: apiHook(documentSymbol.register(options), false),
        findDocumentLinks: apiHook(documentLink.register(options), false),
        findDocumentColors: apiHook(documentColor.register(options), false),
        doRefAutoClose: apiHook(refAutoClose.register(options), false),
        ...createNoStateLanguageService({ typescript }),
        dispose: tsLanguageService.dispose,
    };
    function getShouldUpdateTemplateScript(uri, pos) {
        if (!isInTemplate()) {
            return false;
        }
        update(false); // update tsProjectVersionWithoutTemplate
        if (lastCompletionUpdateVersion !== tsProjectVersionWithoutTemplate) {
            lastCompletionUpdateVersion = tsProjectVersionWithoutTemplate;
            return true;
        }
        return false;
        function isInTemplate() {
            const tsRanges = mapper.ts.to(uri, pos);
            for (const tsRange of tsRanges) {
                if (tsRange.data.vueTag === 'template') {
                    return true;
                }
            }
            const htmlRanges = mapper.html.to(uri, pos);
            if (htmlRanges.length) {
                return true;
            }
            return false;
        }
    }
    function apiHook(api, shouldUpdateTemplateScript = true) {
        const handler = {
            apply: function (target, thisArg, argumentsList) {
                if (typeof shouldUpdateTemplateScript === 'boolean') {
                    update(shouldUpdateTemplateScript);
                }
                else {
                    update(shouldUpdateTemplateScript.apply(null, argumentsList));
                }
                return target.apply(thisArg, argumentsList);
            }
        };
        return new Proxy(api, handler);
    }
    function update(shouldUpdateTemplateScript) {
        var _a;
        const newVueProjectVersion = (_a = vueHost.getProjectVersion) === null || _a === void 0 ? void 0 : _a.call(vueHost);
        if (newVueProjectVersion === undefined || newVueProjectVersion !== vueProjectVersion) {
            let tsFileChanged = false;
            vueProjectVersion = newVueProjectVersion;
            const oldFiles = new Set([...lastScriptVersions.keys()]);
            const newFiles = new Set([...vueHost.getScriptFileNames()]);
            const removes = [];
            const adds = [];
            const updates = [];
            for (const fileName of oldFiles) {
                if (!newFiles.has(fileName)) {
                    if (fileName.endsWith('.vue')) {
                        removes.push(fileName);
                    }
                    else {
                        tsFileChanged = true;
                    }
                    lastScriptVersions.delete(fileName);
                }
            }
            for (const fileName of newFiles) {
                if (!oldFiles.has(fileName)) {
                    if (fileName.endsWith('.vue')) {
                        adds.push(fileName);
                    }
                    else {
                        tsFileChanged = true;
                    }
                    lastScriptVersions.set(fileName, vueHost.getScriptVersion(fileName));
                }
            }
            for (const fileName of oldFiles) {
                if (newFiles.has(fileName)) {
                    const oldVersion = lastScriptVersions.get(fileName);
                    const newVersion = vueHost.getScriptVersion(fileName);
                    if (oldVersion !== newVersion) {
                        if (fileName.endsWith('.vue')) {
                            updates.push(fileName);
                        }
                        else {
                            tsFileChanged = true;
                        }
                        lastScriptVersions.set(fileName, newVersion);
                    }
                }
            }
            if (tsFileChanged) {
                updateTsProject(false);
                updates.length = 0;
                for (const fileName of oldFiles) {
                    if (newFiles.has(fileName)) {
                        if (fileName.endsWith('.vue')) {
                            updates.push(fileName);
                        }
                    }
                }
            }
            const finalUpdates = adds.concat(updates);
            if (removes.length) {
                unsetSourceFiles(removes.map(shared_1.fsPathToUri));
            }
            if (finalUpdates.length) {
                updateSourceFiles(finalUpdates.map(shared_1.fsPathToUri), shouldUpdateTemplateScript);
            }
        }
        else if (shouldUpdateTemplateScript && templateScriptUpdateUris.size) {
            updateSourceFiles([], shouldUpdateTemplateScript);
        }
    }
    function createTsLanguageServiceHost() {
        var _a, _b;
        const scriptSnapshots = new Map();
        // @ts-ignore
        const importSuggestionsCache = (_b = (_a = typescript.Completions) === null || _a === void 0 ? void 0 : _a.createImportSuggestionsForFileCache) === null || _b === void 0 ? void 0 : _b.call(_a);
        const tsHost = {
            ...vueHost,
            fileExists: vueHost.fileExists
                ? fileName => {
                    if (fileName.endsWith('.vue.ts')) {
                        fileName = upath.trimExt(fileName);
                    }
                    return vueHost.fileExists(fileName);
                }
                : undefined,
            getProjectVersion: () => {
                var _a;
                reactivity_1.pauseTracking();
                const version = ((_a = vueHost.getProjectVersion) === null || _a === void 0 ? void 0 : _a.call(vueHost)) + ':' + tsProjectVersion.toString();
                reactivity_1.resetTracking();
                return version;
            },
            getScriptFileNames,
            getScriptVersion,
            getScriptSnapshot,
            readDirectory: (path, extensions, exclude, include, depth) => {
                var _a, _b;
                const result = (_b = (_a = vueHost.readDirectory) === null || _a === void 0 ? void 0 : _a.call(vueHost, path, extensions, exclude, include, depth)) !== null && _b !== void 0 ? _b : [];
                for (const [_, sourceFile] of sourceFiles) {
                    const vuePath = shared_1.uriToFsPath(sourceFile.uri);
                    const vuePath2 = upath.join(path, upath.basename(vuePath));
                    if (upath.relative(path.toLowerCase(), vuePath.toLowerCase()).startsWith('..')) {
                        continue;
                    }
                    if (!depth && vuePath.toLowerCase() === vuePath2.toLowerCase()) {
                        result.push(vuePath2);
                    }
                    else if (depth) {
                        result.push(vuePath2); // TODO: depth num
                    }
                }
                return result;
            },
            // @ts-ignore
            // TODO: crash on 'addListener' from 'node:process', reuse because TS has same problem
            getImportSuggestionsCache: () => importSuggestionsCache,
        };
        return tsHost;
        function getScriptFileNames() {
            const tsFileNames = [];
            tsFileNames.push(shared_1.uriToFsPath(globalDoc.uri));
            for (const fileName of vueHost.getScriptFileNames()) {
                const uri = shared_1.fsPathToUri(fileName);
                const sourceFile = sourceFiles.get(uri);
                if (sourceFile) {
                    for (const [uri] of sourceFile.getTsDocuments()) {
                        tsFileNames.push(shared_1.uriToFsPath(uri)); // virtual .ts
                    }
                }
                if (isTsPlugin) {
                    tsFileNames.push(fileName); // .vue + .ts
                }
                else if (!sourceFile && !fileName.endsWith('.vue')) {
                    tsFileNames.push(fileName); // .ts
                }
            }
            return tsFileNames;
        }
        function getScriptVersion(fileName) {
            const uri = shared_1.fsPathToUri(fileName);
            if (uri === globalDoc.uri) {
                return globalDoc.version.toString();
            }
            for (const [_, sourceFile] of sourceFiles) {
                const doc = sourceFile.getTsDocuments().get(uri);
                if (doc) {
                    return doc.version.toString();
                }
            }
            return vueHost.getScriptVersion(fileName);
        }
        function getScriptSnapshot(fileName) {
            const version = getScriptVersion(fileName);
            const cache = scriptSnapshots.get(fileName);
            if (cache && cache[0] === version) {
                return cache[1];
            }
            const uri = shared_1.fsPathToUri(fileName);
            if (uri === globalDoc.uri) {
                const text = globalDoc.getText();
                const snapshot = typescript.ScriptSnapshot.fromString(text);
                scriptSnapshots.set(fileName, [version, snapshot]);
                return snapshot;
            }
            for (const [_, sourceFile] of sourceFiles) {
                const doc = sourceFile.getTsDocuments().get(uri);
                if (doc) {
                    const text = doc.getText();
                    const snapshot = typescript.ScriptSnapshot.fromString(text);
                    scriptSnapshots.set(fileName, [version, snapshot]);
                    return snapshot;
                }
            }
            let tsScript = vueHost.getScriptSnapshot(fileName);
            if (tsScript) {
                scriptSnapshots.set(fileName, [version, tsScript]);
                return tsScript;
            }
        }
    }
    function getTextDocument(uri) {
        const fileName = shared_1.uriToFsPath(uri);
        const version = Number(vueHost.getScriptVersion(fileName));
        if (!documents.has(uri) || documents.get(uri).version !== version) {
            const scriptSnapshot = vueHost.getScriptSnapshot(fileName);
            if (scriptSnapshot) {
                const scriptText = scriptSnapshot.getText(0, scriptSnapshot.getLength());
                const document = vscode_languageserver_textdocument_1.TextDocument.create(uri, uri.endsWith('.vue') ? 'vue' : 'typescript', version, scriptText);
                documents.set(uri, document);
            }
        }
        if (documents.has(uri)) {
            return documents.get(uri);
        }
        return tsLanguageService.getTextDocument(uri);
    }
    function getSourceFile(uri) {
        return sourceFiles.get(uri);
    }
    function getAllSourceFiles() {
        return [...sourceFiles.values()];
    }
    function updateSourceFiles(uris, shouldUpdateTemplateScript) {
        var _a;
        let vueScriptsUpdated = false;
        let vueTemplateScriptUpdated = false;
        let updateNums = uris.length;
        let currentNums = 0;
        const _shouldSendUpdate = shouldSendUpdate;
        if (shouldUpdateTemplateScript) {
            shouldSendUpdate = false;
        }
        for (const uri of uris) {
            if (_shouldSendUpdate) {
                onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate(currentNums / updateNums);
            }
            currentNums++;
            const sourceFile = sourceFiles.get(uri);
            const doc = getTextDocument(uri);
            if (!doc)
                continue;
            if (!sourceFile) {
                sourceFiles.set(uri, sourceFile_1.createSourceFile(doc, tsLanguageService, typescript, 'api'));
                vueScriptsUpdated = true;
            }
            else {
                const updates = sourceFile.update(doc);
                if (updates.scriptUpdated) {
                    vueScriptsUpdated = true;
                }
                if (updates.templateScriptUpdated) {
                    vueTemplateScriptUpdated = true;
                }
            }
            templateScriptUpdateUris.add(uri);
        }
        if (shouldUpdateTemplateScript) {
            updateNums += templateScriptUpdateUris.size;
        }
        if (vueScriptsUpdated) {
            updateTsProject(false);
        }
        if (shouldUpdateTemplateScript) {
            for (const uri of templateScriptUpdateUris) {
                if (_shouldSendUpdate) {
                    onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate(currentNums / updateNums);
                }
                currentNums++;
                if ((_a = sourceFiles.get(uri)) === null || _a === void 0 ? void 0 : _a.updateTemplateScript()) {
                    vueTemplateScriptUpdated = true;
                }
            }
            templateScriptUpdateUris.clear();
        }
        if (vueTemplateScriptUpdated) {
            updateTsProject(true);
        }
        if (_shouldSendUpdate) {
            onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate(1);
        }
    }
    function unsetSourceFiles(uris) {
        let updated = false;
        for (const uri of uris) {
            if (sourceFiles.delete(uri)) {
                updated = true;
            }
        }
        if (updated) {
            updateTsProject(false);
        }
    }
    function updateTsProject(isTemplateUpdate) {
        tsProjectVersion++;
        if (!isTemplateUpdate) {
            tsProjectVersionWithoutTemplate++;
        }
    }
}
exports.createLanguageService = createLanguageService;
