// @volar/vscode-server/out/instances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setNoStateLs = exports.setServicesManager = exports.noStateLs = exports.servicesManager = exports.allFilesReg = exports.vueFileReg = exports.vueDocReg = exports.documents = exports.connection = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const node_1 = require("vscode-languageserver/node");
exports.connection = node_1.createConnection(node_1.ProposedFeatures.all);
exports.documents = new node_1.TextDocuments(vscode_languageserver_textdocument_1.TextDocument);
exports.vueDocReg = {
    documentSelector: [
        { language: 'vue' },
    ],
};
exports.vueFileReg = {
    documentSelector: [
        { scheme: 'file', language: 'vue' },
    ],
};
exports.allFilesReg = {
    documentSelector: [
        { scheme: 'file', language: 'vue' },
        { scheme: 'file', language: 'javascript' },
        { scheme: 'file', language: 'typescript' },
        { scheme: 'file', language: 'javascriptreact' },
        { scheme: 'file', language: 'typescriptreact' },
    ],
};
function setServicesManager(_servicesManager) {
    exports.servicesManager = _servicesManager;
}
exports.setServicesManager = setServicesManager;
function setNoStateLs(_noStateLs) {
    exports.noStateLs = _noStateLs;
}
exports.setNoStateLs = setNoStateLs;
